### PR TITLE
Improve `DioError`s

### PR DESCRIPTION
**Pull request by ueman**
NOTE: This goes against the `develop-5.0` branch, since it's a breaking change.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dartlang.org/packages/dio)
- [x] I have searched for a similar pull request in the [project](https://github.com/flutterchina/dio/pulls) and found none
- [x] I have updated this branch with the latest `develop` to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I am adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests and they pass

### Pull Request Description

This significantly improves the inner stacktraces of the `DioError`.

### CODE REVIEW

**Comment by jgoyvaerts** [comment]
Good idea! Can you include an example of a stacktrace message before and after your PR so people can see the difference?

**Comment by ueman** [comment]
> Good idea! Can you include an example of a stacktrace message before and after your PR so people can see the difference?

I'm still thinking about how to produce those cases, in order to test it. Do you have any ideas?

**Comment by kuhnroyal** [comment]
Just a reminder too myself: Consider raising to Dart 2.16 and using `Error.rethrowWithStackTrace`.

**Comment by ueman** [comment]
@kuhnroyal Do you mind sharing your test cases? Then I can go through them.

**Comment by kuhnroyal** [comment]
It's nothing special, I only tested:

1. 404 response from a random URL (this has no inner error as the response is received)
2. Something like `post<void>(url, data: ArgumentError())` - this throws a serialization error an has an inner error/stacktrace

File: dio/lib/src/adapters/browser_adapter.dart [hunk 1]
```diff
@@ -151,11 +149,12 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
     xhr.onError.first.then((_) {
       // Unfortunately, the underlying XMLHttpRequest API doesn't expose any
       // specific information about the error itself.
+      // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestEventTarget/onerror
```
Comment (AlexV525) [inline dio/lib/src/adapters/browser_adapter.dart:152]:
```suggestion
      // See also: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestEventTarget/onerror
```

File: dio/lib/src/adapters/io_adapter.dart [hunk 2]
```diff
@@ -56,13 +56,18 @@ class DefaultHttpClientAdapter implements HttpClientAdapter {
...
+        rethrow;
       }
-      rethrow;
-    } on TimeoutException {
-      _throwConnectingTimeout();
+      throw DioError.connectionTimeout(
+        requestOptions: options,
+        timeout: options.connectTimeout ??
+            _httpClient.connectionTimeout ??
+            Duration(),
```
Comment (AlexV525) [inline dio/lib/src/adapters/io_adapter.dart:67]: Maybe `Duration.zero`?

File: dio/lib/src/dio_error.dart [hunk 3]
```diff
@@ -2,35 +2,109 @@ import 'options.dart';
 import 'response.dart';
 
 enum DioErrorType {
-  /// It occurs when url is opened timeout.
-  connectTimeout,
+  /// Caused by a connection timeout.
+  connectionTimeout,
```
Comment (AlexV525) [inline dio/lib/src/dio_error.dart:6]: The changes with `DioErrorType` should be considered as a breaking change, which means you'll need to update the migration guide?

**Review by AlexV525 — COMMENTED** [review]
Hi @ueman. The change looks valid overall, with only a few nit-pickings.

Also, there are a few things I'm considering:
- Are you willing to contribute by letting us pick the PR to another community-maintained `dio`? (We'll announce the new repo later once we merge PRs as much as possible.)
- It might be better to add some tests.
- `.DS_Store` can be removed from the PR.

**Comment by ueman** [comment]
> Hi @ueman. The change looks valid overall, with only a few nit-pickings.
> 
> Also, there are a few things I'm considering:
> 
> * Are you willing to contribute by letting us pick the PR to another community-maintained `dio`? (We'll announce the new repo later once we merge PRs as much as possible.)

Sure.

> * It might be better to add some tests.

I agree, any specific suggestions for tests? Just so I don't miss anything, you might think is important.

> * `.DS_Store` can be removed from the PR.

Yes, good point

File: dio/lib/src/adapters/io_adapter.dart (hunk 2 above)
Comment (ueman) [reply dio/lib/src/adapters/io_adapter.dart:67]: Yep, makes sense

Comment (ueman) [reply dio/lib/src/adapters/io_adapter.dart:67]:
```suggestion
            Duration.zero,
```

File: dio/lib/src/dio_error.dart (hunk 3 above)
Comment (ueman) [reply dio/lib/src/dio_error.dart:6]: Yes, this is a breaking change

**Comment by AlexV525** [comment]
> Sure.

Thanks!

> I agree, any specific suggestions for tests? Just so I don't miss anything, you might think is important.

IMO tests should be covered by checking if requests throw corresponding exceptions in condition.

File: dio/lib/src/dio_error.dart (hunk 3 above)
Comment (ueman) [reply dio/lib/src/dio_error.dart:6]: I don't think I need to add anything to the migration guide. From a user's point of view, nothing changes. The user can still catch exceptions the same way as before. All breaking changes are done from within the package in this PR.

File: dio/lib/src/dio_mixin.dart [hunk 4]
```diff
@@ -630,6 +636,11 @@ abstract class DioMixin implements Dio {
         }
       }
 
+      StackTrace? stackTrace;
+      if (maybeStackTrace is StackTrace?) {
```
Comment (AlexV525) [inline dio/lib/src/dio_mixin.dart:640]: `maybeStackTrace` will always be `StackTrace`. The condition can be `maybeStackTrace != StackTrace.empty`.

File: dio/CHANGELOG.md [hunk 5]
```diff
@@ -1,17 +1,23 @@
-# Unreleased
-- require Dart `2.12.1` which fixes exception handling for secure socket connections (https://github.com/dart-lang/sdk/issues/45214)  
-- `ResponseBody.statusCode` is now non-nullable
-- Only delete file if it exists when downloading.
-
 # Unreleased 5.0
-- add option to instantiate a `HttpClientAdapter`, which is platform independent
 
-# 4.0.5-beta1
+- Add option to instantiate a `HttpClientAdapter`, which is platform independent
```
Comment (AlexV525) [inline dio/CHANGELOG.md:3]: Seems like we need to change the base branch to `develop`.

File: dio/lib/src/adapters/browser_adapter.dart [hunk 6]
```diff
@@ -74,17 +74,17 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
...
-            completer.completeError(
-              DioError(
-                requestOptions: options,
-                error: 'Connecting timed out [${options.connectTimeout}ms]',
-                type: DioErrorType.connectTimeout,
-              ),
-              StackTrace.current,
-            );
-            xhr.abort();
+          if (haveSent) {
```
Comment (AlexV525) [inline dio/lib/src/adapters/browser_adapter.dart:77]: This will be fix by another PR.

File: dio/lib/src/dio_error.dart [hunk 7]
```diff
@@ -2,35 +2,109 @@ import 'options.dart';
...
 
-  /// When the server response, but with a incorrect status, such as 404, 503...
-  response,
+  /// The [DioError] was caused by an incorrect status code as configured by
+  /// [ValidateStatus].
+  badResponse,
 
   /// When the request is cancelled, dio will throw a error with this type.
-  cancel,
+  requestCancelled,
```
Comment (AlexV525) [inline dio/lib/src/dio_error.dart:19]: Users usually ignore canceled requests since they are meant to do it, which means they'll rely on the en. Maybe we can keep the naming at least for the least breaking.

**Review by AlexV525 — COMMENTED** [review]
Given that we've already merged a couple PRs in the fork, I would suggest that we can submit this PR to the fork directly.

**Review by AlexV525 — COMMENTED** [review]
I would now suggest submitting the PR to the fork directly. Also, the target branch should be `main`.

**Comment by ueman** [comment]
> I would now suggest submitting the PR to the fork directly. Also, the target branch should be `main`.

I don't have the permissions to do that

**Comment by AlexV525** [comment]
> I don't have the permissions to do that

Sorry. It should be granted now.

**Comment by AlexV525** [comment]
Hi everyone! We've made our hardfork repo public and published a new version of `dio`, named `diox`.
The new package contains the PR of the fix.
Please refer to https://pub.dev/packages/diox/versions/5.0.0-dev.1 to use the fork.
You can also see why we're working for a hardfork at https://github.com/cfug/diox/issues/29 and https://github.com/flutterchina/dio/issues/1607.

**Comment by kuhnroyal** [comment]
All the server response errors have no stackTrace at all. It is impossible to tell where the error originated.
